### PR TITLE
ENH: Set type (formerly ParamList)

### PR DESCRIPTION
**Pull request by kszucs**
This should be applied on top of #1292 and #1300 one they are got merged.
It also requires to generalize `ListExpr` and `ValueList` a bit.

### CODE REVIEW

**Comment by cpcloud** [comment]
@kszucs Looks like there are some merge conflicts here.

**Comment by kszucs** [comment]
@cpcloud conflict resolved

File: ibis/expr/tests/test_value_exprs.py [hunk 1]
```diff
@@ -1076,6 +1076,13 @@ def test_time_invalid_compare_on_py2():
     assert not result
 
 
+def test_scalar_parameter_list():
+    value = ibis.param([dt.int64, dt.int32])
+
+    assert isinstance(value.op(), ir.ParamList)
+    assert value.type().equals(dt.int64)
```
Comment (cpcloud) [inline ibis/expr/tests/test_value_exprs.py]: This interface feels a bit confusing to me. Why am I allowed to pass multiple types in a list?

File: ibis/expr/types.py [hunk 2]
```diff
@@ -612,6 +615,11 @@ def param(type, name=None):
     """
     if name is None:
         name = _parameter_name()
+
+    if isinstance(type, (tuple, list)):
+        dtype = dt.highest_precedence(map(dt.dtype, type))
+        return ParamList([param(dtype)]).to_expr()
```
Comment (cpcloud) [inline ibis/expr/types.py]: Should this also have a `.name(name)` call after to `to_expr()` call?

File: ibis/expr/tests/test_value_exprs.py (hunk 1 above)
Comment (kszucs) [reply ibis/expr/tests/test_value_exprs.py]: Agree, but I didn't have a better API in mind. 

Actually we might use `ibis.param(dt.Array(dt.int64))` instead. 
I'm not sure which one could better fit with `IN` clauses.

I'm more than open for creating a better API.

Comment (kszucs) [reply ibis/expr/tests/test_value_exprs.py]: @cpcloud ping

Comment (cpcloud) [reply ibis/expr/tests/test_value_exprs.py]: Hm, I think we need something like this, but I'm wondering if something like a `Set` type is the way to go long-term. Is this feature blocking you at the moment?

`Contains` would now look a bit different with a `Set` type.

```python
class Contains:
    input_type = [rules.value, rules.set(rules.value)]
```

then for params we have `ibis.param({dt.int64})` as the API (could also allow `list`s there as well).

We could probably even get rid of `ValueList` altogether and just have `Set`s and `array` types.

**Comment by kszucs** [comment]
Blocking, but your approach is way better. I'm implementing this tonight.

File: ibis/expr/tests/test_value_exprs.py [hunk 3]
```diff
@@ -254,16 +253,16 @@ def test_negate_isin():
...
 
-    expr = val.isin(options)
-    assert isinstance(expr, ir.BooleanColumn)
+#     expr = val.isin(options)
+#     assert isinstance(expr, ir.BooleanColumn)
 
-    not_expr = val.notin(options)
-    assert isinstance(not_expr, ir.BooleanColumn)
+#     not_expr = val.notin(options)
+#     assert isinstance(not_expr, ir.BooleanColumn)
```
Comment (kszucs) [inline ibis/expr/tests/test_value_exprs.py]: We need another operation to support this.

File: ibis/clickhouse/tests/test_operators.py [hunk 4]
```diff
@@ -132,12 +132,15 @@ def test_string_temporal_compare_between_datetimes(con, left, right):
 
 
 def test_field_in_literals(con, alltypes, translate):
-    expr = alltypes.string_col.isin(['foo', 'bar', 'baz'])
-    assert translate(expr) == "`string_col` IN ('foo', 'bar', 'baz')"
+    foobar = set(['foo', 'bar', 'baz'])
+    foobar_ = str(tuple(foobar))
+
+    expr = alltypes.string_col.isin(foobar)
+    assert translate(expr) == "`string_col` IN {}".format(foobar_)
```
Comment (kszucs) [inline ibis/clickhouse/tests/test_operators.py]: Doesn't keep order of elements.

File: ibis/expr/datatypes.py [hunk 5]
```diff
@@ -1004,6 +1034,20 @@ def from_string(value):
     return TypeParser(value).parse()
 
 
+@dtype.register(list)
+def from_list(values):
+    if not values:
+        return Array(null)
+    return Array(highest_precedence(map(dtype, values)))
```
Comment (kszucs) [inline ibis/expr/datatypes.py]: To allow `dt.dtype([dt.int8])` syntax.

File: ibis/expr/types.py [hunk 6]
```diff
@@ -1434,7 +1446,10 @@ def literal(value, type=None):
     if hasattr(value, 'op') and isinstance(value.op(), Literal):
         return value
 
-    dtype = infer_literal_type(value)
+    if value is null:
```
Comment (kszucs) [inline ibis/expr/types.py]: Why is `null` a callable? To ensure immutability of `_NULL`?

Should I register a to `dt.infer` with `FunctionType`?

File: ibis/impala/tests/test_exprs.py [hunk 7]
```diff
@@ -732,36 +732,40 @@ def setUp(self):
...
         self._check_expr_cases(cases)
 
+#     def test_literal_in_list(self):
+#         cases = [
+#             (L(2).isin([self.table.a, self.table.b, self.table.c]),
+#              '2 IN (`a`, `b`, `c`)'),
+#             (L(2).notin([self.table.a, self.table.b, self.table.c]),
+#              '2 NOT IN (`a`, `b`, `c`)')
+#         ]
+#         self._check_expr_cases(cases)
```
Comment (kszucs) [inline ibis/impala/tests/test_exprs.py]: Requires an additional operation.

File: ibis/impala/tests/test_exprs.py [hunk 8]
```diff
@@ -1101,7 +1105,7 @@ def test_builtins_1(self):
             5 / i4.nullif(0),
             5 / d.nullif(0),
 
-            api.literal(5).isin([i1, i4, d]),
+            # api.literal(5).isin([i1, i4, d]),
```
Comment (kszucs) [inline ibis/impala/tests/test_exprs.py]: Additional op. How should the signature look like?

**Comment by kszucs** [comment]
@cpcloud I'm not sure that we can get rid of `ValueExpr` without feature degradation.

Which backends support `literal(5).isin([i1, i4, d])` like expressions?

**Comment by cpcloud** [comment]
To be clear, I suggested getting rid of ValueList, not ValueExpr. 

Isin is supported by all backends, if it's not that's a bug.

**Comment by cpcloud** [comment]
The new isin should take a value and a set of values.

**Comment by kszucs** [comment]
@cpcloud I ment `ValueList` :)

**Comment by kszucs** [comment]
Actually I'm totally lost between the following classes:

- `ExpressionList(Node)`
- `ExprList(Expr)`
- `ListExpr(ColumnExpr, AnyValue)`
- `ValueList(ValueOp)`

**Comment by kszucs** [comment]
@cpcloud I think we should schedule this to 0.14

**Comment by cpcloud** [comment]
@kszucs Looks like the rebase is going to be pretty rough here. Let me know when this is ready to be merged.

**Comment by kszucs** [comment]
@cpcloud I've rather reimplemented in #1420, closing in favor of that.

**Pull request closed without merging**